### ISSUE TITLE
Delay when closing nvim with rust_analyzer active, first call to stop() ignored but fixes delay

### ISSUE THREAD

**Issue by nerosnm** [neovim/neovim#12478]
This issue was originally posted as neovim/nvim-lsp#275, but all the info from there is summarised below.

- `nvim --version`:

    ```
    NVIM v0.5.0-nightly-83-ge78658348
    Build type: Release
    LuaJIT 2.0.5
    Compilation: /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNDEBUG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/tmp/neovim-20200611-6709-f95l5p/build/config -I/tmp/neovim-20200611-6709-f95l5p/src -I/usr/local/include -I/tmp/neovim-20200611-6709-f95l5p/deps-build/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include -I/usr/local/opt/gettext/include -I/tmp/neovim-20200611-6709-f95l5p/build/src/nvim/auto -I/tmp/neovim-20200611-6709-f95l5p/build/include
    Compiled by soren@Fafnir.local

    Features: +acl +iconv +tui
    See ":help feature-compile"

       system vimrc file: "$VIM/sysinit.vim"
      fall-back for $VIM: "/usr/local/Cellar/neovim/HEAD-e786583/share/nvim"

    Run :checkhealth for more info
    ```
- `vim -u DEFAULTS` behaves differently? N/A, involves nvim-lsp
- Operating system/version: macOS 10.15.5 (19F101)
- Terminal name/version: alacritty 0.4.2 (f68de37)
- `$TERM`: screen-256color

### How to reproduce the problem from neovim startup

Install and configure nvim-lsp through nvim-plug with rust-analyzer by putting the following in `init.vim`:

```vim
if empty(glob('~/.local/share/nvim/site/autoload/plug.vim'))
  silent !curl -fLo ~/.local/share/nvim/site/autoload/plug.vim --create-dirs
    \ https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
  au VimEnter * PlugInstall --sync | source $MYVIMRC
endif

call plug#begin('~/.local/share/nvim/plugged')
Plug 'neovim/nvim-lsp'
call plug#end()

lua << EOF
nvim_lsp = require'nvim_lsp'

nvim_lsp.rust_analyzer.setup {
    settings = {
        ['rust-analyzer'] = {
            completion = {
                addCallArgumentSnippets = false,
                addCallParenthesis = false,
            },
        },
    },
}
EOF
```

Open nvim, run `:PlugInstall`, and then close nvim and reopen it on a file in a Rust project, and then quit nvim with `:q`.

### Actual behaviour

There is a delay of between 0.5-1s once `:q` is run, before `nvim` closes.

If `:lua vim.lsp.stop_client(vim.lsp.get_active_clients())` is run, then `:q` subsequently causes nvim to quit instantaneously. However, that command _doesn't_ actually stop the client, and it doesn't seem like it's stopping a duplicate client, as evidenced below.

At startup:

- `:lua print(#vim.lsp.get_active_clients())` prints 1
- `:lua print(vim.lsp.get_active_clients()[1].is_stopped())` prints `false`
- There is a 0.5-1s delay when running `:q` before `nvim` closes.

However, after calling `:lua vim.lsp.get_active_clients()[1].stop()` once:

- `:lua print(#vim.lsp.get_active_clients())` still prints 1
- `:lua print(vim.lsp.get_active_clients()[1].is_stopped())` still prints `false`
- Completions and error message display in the buffer still work.
- `:q` causes `nvim` to close instantaneously.

After making a second call to stop the client:

- `:lua print(#vim.lsp.get_active_clients())` prints 0
- `:lua print(vim.lsp.get_active_clients()[1].is_stopped())` errors because `vim.lsp.get_active_clients()[1]` is `nil`.
- Completions and error message display in the buffer do not work.

### Expected behaviour

`nvim` closes instantaneously, and a single call to `:lua vim.lsp.get_active_clients()[1].stop()` stops the rust-analyzer client rather than this requiring two calls.

**Comment by nerosnm** [comment]
@nacknime-official, you mentioned that you get the same delay when closing `nvim` with pyls. Could you check whether you get the same behaviour when asking a client to stop, as above?

**Comment by gbrlsnchs** [comment]
I just want to add that this is in fact abnormal behavior. I tested `rust-analyzer` with another editor (Kakoune using `kak-lsp`) and there is no such issue there. Thanks @nerosnm for opening this issue!

**Comment by ludwj61** [comment]
I just want to add I also get this issue with all lsp servers on nightly. It doesn't occur when using lsp servers with coc instead of built-in.

**Comment by fsouza** [comment]
I started seeing this too (with all servers). Will spend some time over the next couple of days trying to poke at it to see if I can provide a better issue report.

**Comment by nerosnm** [comment]
Seems like I'm now getting a delay not only when closing `nvim` (and so stopping the client) but also every time I open a new file (including the first time I open `nvim`). If I `set nohidden`, that means I get a pause of 0.5-1s every time I switch files.

**Comment by nerosnm** [comment]
Disregard my last comment, I had inadvertently removed `set shell=/bin/bash` from my init file and apparently `fish` causes those big delays. I am still getting a slight delay when closing `nvim` though.

**Comment by fsouza** [comment]
> I started seeing this too (with all servers). Will spend some time over the next couple of days trying to poke at it to see if I can provide a better issue report.

This seems fine now here too. So nothing to dig on 🙈

**Comment by dimbleby** [comment]
Delay when closing neovim when is still a thing for me!  

I've just this last week tried switching to that builtin LSP client, and the delay when exiting is quite noticeable (which is why I went and found and subscribed to this issue).

**Comment by mjlbach** [comment]
I don't have this issue with rust-analyzer, so it would be helpful if you could pin down a reproduction

**Comment by nerosnm** [comment]
I _am_ experiencing this with rust-analyzer, but as far as I can tell, only on Darwin. My NixOS machine doesn't have this problem.

**Comment by mjlbach** [comment]
@nerosnm Can you confirm the versions of rust-analyzer are the same?

**Comment by fsouza** [comment]
(to be clear, I was not seeing this issue with rust-analyzer, but with other servers. Can't repro anymore, but I was also on darwin)

**Comment by dimbleby** [comment]
> pin down a reproduction

I'd love for a reproduction where this _doesn't_ happen!  Repro is just: use the built-in client and rust-analyzer, quit neovim.

For me at least, I think it's as simple as: the built-in client gracefully shuts down servers before exiting.  And at least for some servers - like others, I find this most conspicuous with rust-analyzer - this takes a noticeable amount of time.

I suspect that other clients just don't bother to shut down the servers at exit.  After all, the language server is a child of the neovim process, so it'll die when neovim dies.  To confirm: if I comment out the body of `lsp._vim_exit_handler()`, closing neovim is once again fast - and rust-analyzer exits anyway.

**Comment by mjlbach** [comment]
I'm guessing this is what y'all are noticing? 

https://user-images.githubusercontent.com/13316262/109574971-b7656480-7aa5-11eb-9dee-2425a4f32478.mov

For me, it seems more like 100-200 ms.

**Comment by dimbleby** [comment]
Yeah, definitely not much; and perhaps not much to get excited about at all.

But having just made the switch my fingers aren't yet used to it and I find myself occasionally missing a character or two of the next command after exiting!

**Comment by mjlbach** [comment]
@tjdevries I guess we don't *have* to shut down servers elegantly. I worry about things though like terminating the server in the middle of a formatting/write operation

**Comment by fsouza** [comment]
@mjlbach you could send the signal and not wait for the server to die, but that has the risk of leaving stuff behind in case a server misbehaves x)

**Comment by mjlbach** [comment]
Closed by https://github.com/neovim/neovim/pull/14180

**Comment by mjlbach** [comment]
I'm reverting #14180 because it closes neovim without waiting to send pkill to active language servers, so this issue remains open until we find a good workaround

**Comment by dimbleby** [comment]
re #14180 causing orphaned servers - it might be reasonable to consider that such cases are bugs in the server and should be treated as such.  Eg jedi-language-server already has an open issue effectively saying that it should exit when its client exits (and I will submit an MR to do that.)

I wonder whether the same could reasonably be said for the whole force-stop / graceful shutdown thing.  If there are servers out there that don't follow the protocol re shutdown, let users raise bugs against those servers: it's going too far for clients to start throwing SIGTERMs around.

ie I think I'm suggesting simplification:

- client.stop just performs the "graceful" shutdown path
- _vim_exit_handler just stops clients (gracefully), and doesn't bother to wait for them

Admittedly if you had have done this, I would have encountered jedi-language-server not exiting and raised #14428 anyway.  But on reflection I think it would have been legit for the answer to be: fix the server, please.

**Comment by Thiago4532** [comment]
Has anyone found the solution to this issue? I am having the same problem using ccls.

**Comment by mjlbach** [comment]
> re #14180 causing orphaned servers - it might be reasonable to consider that such cases are bugs in the server and should be treated as such. Eg jedi-language-server already has an open issue effectively saying that it should exit when its client exits (and I will submit an MR to do that.)


I think it's rather the opposite, servers that shut down slowly in response to the "graceful" shutdown path are the ones that are buggy.

@Thiago4532 no, you can see the linked PR, the reverting of the PR, and the discussion as to why on the associated issues. If you want to discuss possible solutions with the community (I don't think there is a good solution other than sig-killing language servers or reducing the vim.wait timeout), discourse is the venue.

**Comment by Thiago4532** [comment]
Should I submit the PR to pkill language servers instead of waiting them to finish? I don't know if this solution can break some language servers but the only way to know is to test.

**Comment by dimbleby** [comment]
> I think it's rather the opposite, servers that shut down slowly in response to the "graceful" shutdown path are the ones that are buggy.

These things are not opposite, you might reasonably think _both_ that:

- a server that takes a long time to shut down is behaving badly
- a server that continues to run after its client has exited is behaving badly

For what it's worth I don't think that a non-immediate shutdown is buggy, it seems reasonable to me that a server might have state that it wants to persist or other housekeeping to do.

However a server that allows itself to be orphaned _is_ misbehaving: servers ought to allow at least for the possibility of their client crashing.  (I believe this is why the client process ID is passed to the server at initialization.)

Also: I again say that for a client to start sending SIGTERMs (and worse) is going too far: this very clearly not part of the protocol!

I guess I'm just repeating myself, and so I promise not to do it again, but the following still looks satisfactory to me:

- client.stop just performs the "graceful" shutdown path
- _vim_exit_handler just stops clients (gracefully), and doesn't bother to wait for them

**Comment by Thiago4532** [comment]
> > I think it's rather the opposite, servers that shut down slowly in response to the "graceful" shutdown path are the ones that are buggy.
> 
> These things are not opposite, you might reasonably think _both_ that:
> 
>     * a server that takes a long time to shut down is behaving badly
> 
>     * a server that continues to run after its client has exited is behaving badly
> 
> 
> For what it's worth I don't think that a non-immediate shutdown is buggy, it seems reasonable to me that a server might have state that it wants to persist or other housekeeping to do.
> 
> However a server that allows itself to be orphaned _is_ misbehaving: servers ought to allow at least for the possibility of their client crashing. (I believe this is why the client process ID is passed to the server at initialization.)
> 
> Also: I again say that for a client to start sending SIGTERMs (and worse) is going too far: this very clearly not part of the protocol!
> 
> I guess I'm just repeating myself, and so I promise not to do it again, but the following still looks satisfactory to me:
> 
>     * client.stop just performs the "graceful" shutdown path
> 
>     * _vim_exit_handler just stops clients (gracefully), and doesn't bother to wait for them

It is necessary to kill the language servers, especially if they crash and refuse SIGTERM.

**Comment by dimbleby** [comment]
I just don't agree.  Raise bugs against language servers that don't exit when they should.

(Do we actually know of any such?  I hit this with jedi-language-server but, as promised, the MR is done and it doesn't have that bug any more.)

Insisting on seeing servers exit is IMO giving up on resolving the delay that this issue describes, I think it's unlikely that servers can be persuaded to abandon whatever graceful shutdown they are currently doing.  And what's more: sending SIGTERM / SIGKILL risks making a mess of that graceful shutdown.  

(An alternative position - though not the one that I'd prefer - would be that we should just learn to live with the shutdown delay, that's the price you pay for using language servers.)

**Comment by neovim-discourse** [comment]
This issue has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/lag-at-neovim-shutdown-lsp-rust-analyzer-suspected/1221/2

**Comment by dimbleby** [comment]
> I'm not going to have neovim sigkill language servers earlier

That wasn't the proposal.  Sending SIGTERM is already going way beyond what the client ought to be doing IMO, SIGKILL only more so.

>  We also tried only sending the close request, which resulted in many servers not terminating, leading to additional user complaints.

The only such that I know about is #14428, which was opened by ... me!  Clearly I have changed my mind about how such cases should be handled - and also put my money where my mouth is, getting an MR accepted to fix the not-terminating language server.

Closing this is okay by me, I just wouldn't want it to be for the wrong reasons.

**Comment by mjlbach** [comment]
> That wasn't the proposal. Sending SIGTERM is already going way beyond what the client ought to be doing IMO, SIGKILL only more so.

My point was, there are two options here.

1. Send close and be done with it. I tried this, and had even merged https://github.com/neovim/neovim/pull/14180 (which effectively does this, although it had other issues due to the timing of outstanding callbacks being closed before servers) which as far as I know you are done with it
2. Send close, and reduce the time we block closing to wait for the server before sending sigterm or sigkill (I was using the signals flippantly)

We/I tried 1, it's going to lead to orphaned servers, which is worse than an arguably imperceptible delay closing RA (I think it's smaller on better hardware, as it is an RA performance issue).

**Comment by mjlbach** [comment]
> To make clear the sort of thing that I would favour: I am currently experimenting with simply removing this code from neovim:

Yes, that is exactly what I tried, and I observed orphaned language servers. We are on the same page.

**Comment by mjlbach** [comment]
@mfussenegger I'm personally ok with adding an option to flags `force_close_on_exit` that users can disable, and then explicitly it's not our problem if the servers don't close.

**Comment by mfussenegger** [comment]
I'd be in favour of adding a `graceful_exit_timeout` flag or something like that (basically what's currently hardcoded to 500)
Keep the default of 500, and have <= 0 not wait for the client to exit.

Then users can set it to 0 if they don't care about whether the client exits correctly, or they can even increase it for language servers where 500ms might be too low and they want to ensure it stops properly. (Servers like jdtls store workspace files on stop, and that can take more than 500ms)
